dill/session.py

```diff
@@ -213,32 +305,39 @@ def dump_module(
 
     from .settings import settings
     protocol = settings['protocol']
+    if refimported is None:
+        refimported = settings['dump_module']['refimported']
+    if refonfail is None:
+        refonfail = settings['dump_module']['refonfail']
     main = module
     if main is None:
         main = _main_module
     elif isinstance(main, str):
         main = _import_module(main)
     if not isinstance(main, ModuleType):
         raise TypeError("%r is not a module" % main)
-    if hasattr(filename, 'write'):
-        file = filename
-    else:
-        file = open(filename, 'wb')
-    try:
+    original_main = main
+    if refimported:
+        main = _stash_modules(main)
+    with _open(filename, 'wb') as file:
         pickler = Pickler(file, protocol, **kwds)
-        pickler._original_main = main
-        if refimported:
-            main = _stash_modules(main)
+        if main is not original_main:
+            pickler._original_main = original_main
         pickler._main = main     #FIXME: dill.settings are disabled
         pickler._byref = False   # disable pickling by name reference
         pickler._recurse = False # disable pickling recursion for globals
         pickler._session = True  # is best indicator of when pickling a session
         pickler._first_pass = True
-        pickler._main_modified = main is not pickler._original_main
+        if refonfail:
+            pickler._refonfail = True  # False by default
+            pickler._file_seek = getattr(file, 'seek', None)
+            pickler._file_truncate = getattr(file, 'truncate', None)
+            if hasattr(file, 'seekable') and not file.seekable():
+                pickler._file_seek = None
+            if pickler._file_seek is None or pickler._file_truncate is None:
+                raise TypeError("file must have 'tell', 'seek' and 'truncate'"
```

Comment:
Hmm... I'm not sure if I like the doc here.  It's very accurate, however.

Reply:
It's a restriction of the implementation. I couldn't find better words... Even wrapping the stream in a `io.BufferedRandom` object wouldn't work. A plausible workaround for this would be to write the pickle stream to memory (`io.BytesIO`) and write to the file at the end. Except for really big objects and low memory situations, the overhead may be relatively small.

Reply:
Did basically this, with some modifications, and it worked! 🎉